lib/Dialect/HW/HWTypes.cpp

```diff
@@ -311,33 +314,56 @@ Type UnionType::getFieldType(mlir::StringRef fieldName) {
 // ArrayType
 //===----------------------------------------------------------------------===//
 
-Type ArrayType::parse(AsmParser &p) {
-  SmallVector<int64_t, 2> dims;
+template <typename TArray>
+static Type parseArray(AsmParser &p) {
+  if (p.parseLess())
+    return Type();
+
+  Attribute dim;
   Type inner;
-  if (p.parseLess() || p.parseDimensionList(dims, /* allowDynamic */ false) ||
-      parseHWElementType(inner, p) || p.parseGreater())
+  uint64_t dimLiteral;
+  auto int32Type = p.getBuilder().getIntegerType(32);
+
+  if (auto res = p.parseOptionalInteger(dimLiteral); res.hasValue()) {
+    dim = p.getBuilder().getI32IntegerAttr(dimLiteral);
+  } else if (!p.parseOptionalAttribute(dim, int32Type).hasValue())
     return Type();
-  if (dims.size() != 1) {
-    p.emitError(p.getNameLoc(), "hw.array only supports one dimension");
+
+  if (!dim.isa<IntegerAttr, ParamExprAttr, ParamDeclRefAttr>()) {
+    p.emitError(p.getNameLoc(), "unsupported dimension kind in hw.array");
     return Type();
   }
 
+  if (p.parseXInDimensionList() || parseHWElementType(inner, p) ||
+      p.parseGreater())
+    return Type();
+
   auto loc = p.getEncodedSourceLoc(p.getCurrentLocation());
-  if (failed(verify(mlir::detail::getDefaultDiagnosticEmitFn(loc), inner,
-                    dims[0])))
+  if (failed(TArray::verify(mlir::detail::getDefaultDiagnosticEmitFn(loc),
```

Comment:
If you move the construction to the actual parse members, this function doesn't need to be a template function.

Reply:
Modified things slightly, let me know what you think!